src/librustc_mir/transform/simplify_try.rs

```diff
@@ -577,3 +577,131 @@ impl<'tcx> MirPass<'tcx> for SimplifyBranchSame {
+                    if let Some(last) = &bbs[bb_idx].statements.last() {
+                        match &last.kind {
+                            StatementKind::Assign(x) if Some(x.0) == switch_discr.place() => {
+                                match x.1{
+                                    Rvalue::Discriminant(discr_place) => {
+                                        (discr_place, targets, values)
+                                    }
+                                    _ => continue
+                                }
+                            },
+                            _ => continue
+                        }                        
+                    } else {
+                        trace!("NO: discriminant not found that read into switchInt");
+                        continue;
+                    }
+                    
+                },
+                _ => continue,
+            };
+
+            for (target_idx, (bb_idx,bbd)) in targets
+                    .iter()
+                    .map(|idx| (*idx, &bbs[*idx]))
+                    .enumerate() {
+                // we are looking for a basic block of only a single statement; discriminant(place) = VAR_INDEX
+                // this can be replaced with `place = discr_place` if the value of the switch statement matches the VAR_INDEX
+                if bbd.statements.len() != 1 {
+                    trace!("NO: Basic block is not just a single statement; discriminant(place) = VAR_INDEX");
+                    continue;
+                }
+                
+                let stmt = &bbd.statements[0];
+                match &stmt.kind {
+                    StatementKind::SetDiscriminant { place, variant_index } => {
+                        let place_type = place.ty(body, tcx).ty;
+                        let adt = match place_type.kind {
+                            ty::Adt(adt, _) if adt.is_enum() => adt,
+                            _ => {
+                                bug!("set descriminant on adt that is not enum")
+                            }
+                        };
+                        let variant_is_fieldless = adt.variants[*variant_index].fields.is_empty();
+                        
+                        if !variant_is_fieldless{
+                            trace!("NO: variant is not fieldless");
+                            continue; 
+                        } 
+                        let local = place.as_local();
+                        if local.is_none(){
+                            trace!("NO: place for enum is not a local variable");
+                            continue;
+                        } 
+                        let local = local.unwrap();
+                        let discr = adt.discriminant_for_variant(tcx, *variant_index);
+
+                        if body.local_decls()[discr_place.local].ty != body.local_decls()[local].ty {
```

Comment:
I would have thought that this check would guard against the type mismatch errors seen in the PR message. But it does not seem to

Reply:
Sooo... the error happens because I forgot a continue to actually skip optimizing...